File: go/2007.go
package main

func findOriginalArray(changed []int) []int {
	cnt := map[int]int{}
	for _, num := range changed {
		cnt[num]++
	}
	cnt0 := cnt[0]
	if cnt0%2 == 1 {
		return nil
	}
	delete(cnt, 0)
	ans := make([]int, cnt0/2, len(changed)/2)
	done := map[int]bool{}
	for x := range cnt {
		if done[x] || x%2 == 0 && cnt[x/2] > 0 {
			continue
		}
		for cnt[x] > 0 {
			cntX := cnt[x]
			if cntX > cnt[2*x] {
				return nil
			}
			for i := 0; i < cntX; i++ {
				ans = append(ans, x)
			}
			done[x] = true
			cnt[2*x] -= cntX
			if cnt[2*x] == 0 {
				done[2*x] = true
				x *= 4
			} else {
				x *= 2
			}
		}
	}
	return ans
}

File: python/2007.py
from collections import Counter
from typing import List

class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        cnt = Counter(changed)
        cnt0 = cnt.pop(0, 0)
        if cnt0 % 2:
            return []
        ans = [0] * (cnt0 // 2)
        done = set()
        for x in cnt:
            if x in done or x % 2 == 0 and x // 2 in cnt:
                continue
            while x in cnt:
                cnt_x = cnt[x]
                if cnt_x > cnt[2 * x]:
                    return []
                ans.extend([x] * cnt_x)
                done.add(x)
                cnt[2 * x] -= cnt_x
                if cnt[2 * x] == 0:
                    done.add(2*x)
                    x*=4
                else:
                    x *= 2
        return ans